source/MRMesh/MRMeshComponents.cpp

```diff
@@ -299,6 +299,44 @@ std::vector<FaceBitSet> getAllComponents( const MeshPart& meshPart, FaceIncidenc
     return res;
 }
 
+std::vector<FaceBitSet> getAllFlatComponents( const MeshPart& meshPart, float zTolerance )
+{
+    if ( zTolerance < 0 )
+        return {};
+
+    const auto& mesh = meshPart.mesh;
+    const FaceBitSet& region = mesh.topology.getFaceIds( meshPart.region );
+    const auto numFaces = region.find_last() + 1;
+
+    Vector<Vector2f, FaceId> zRanges( numFaces );
+    BitSetParallelFor(region, [&] (FaceId f)
+    {
+            Vector3f triPoints[3];
+            meshPart.mesh.getTriPoints( f, triPoints );
+
+            zRanges[f].x = std::min( { triPoints[0].z, triPoints[1].z, triPoints[2].z } );
+            zRanges[f].y = std::max( { triPoints[0].z, triPoints[1].z, triPoints[2].z } );
+    } );
+
+    return getAllComponents( meshPart, MeshComponents::PerEdge, [&]( UndirectedEdgeId ue ) -> bool
+    {
+        auto f0 = mesh.topology.left( ue );
+        auto f1 = mesh.topology.right( ue );
+
+        const float zMax = std::max( zRanges[f0].y, zRanges[f1].y );
+        const float zMin = std::min( zRanges[f0].x, zRanges[f1].x );
+
+        if ( zMax - zMin < zTolerance )
+        {
+            zRanges[f0].x = zRanges[f1].x = zMin;
+            zRanges[f0].y = zRanges[f1].y = zMax;
```

Comment:
It is wrong. You change the values of only two faces in the region that can include thousands of faces